[[ AndroidListeners ]] Add support for Android listener callbacks in LCB

livecodeali opened the pull request:
This patch adds a binding string variant for Android which allows the user to create an interface proxy - that is an instance of a generic Proxy class for a given interface.
This effectively allows LCB handlers to be registered as the targets for java interface callbacks, such as event listeners.

Either a Handler or an Array can be registered for a given listener- if it is a Handler, the specified listener should only have one available callback. If the listener has multiple callbacks, an array can be used to assign handlers to each. Each key in the array must match the name of a callback in the listener.

We do this by creating a class `LCBInvocationHandler` derived from the InvocationHandler interface, which can create a Proxy object for any given interface. When listener callbacks are invoked on the specified interface, the invocation handler's 'invoke' method is invoked. This calls back into the engine which tries to invoke the selected LCB handler with the jobject arguments passed to the native invoke method.

livecodeali committed:
[[ JavaFFI ]] Add auto class for Java Object ValueRef

livecodeali committed:
[[ JavaFFI ]] Add conversion from jobjectArray to Proper List

livecodeali committed:
[[ AndroidListeners ]] Add method to invoke chosen lcb handler from listener callback

livecodeali committed:
[[ JavaFFI ]] Cache android class loader
The class loader for the android Engine Activity is needed to find
LiveCode defined classes from native code. Cache the class loader
and add an accessor for use with Java FFI binding string resolution
and method call.

livecodeali committed:
[[ JavaFFI ]] Resolve 'com.runrev.android' classes using cached class loader

livecodeali committed:
[[ AndroidListeners ]] Add native method for calling back into engine from android listener invoke

livecodeali committed:
[[ AndroidListeners ]] Add InvocationHandler implementation
The LCBInvocationHandler class is used to create a Proxy for
listener interfaces - when the listener action happens, the
invoke method is triggered with the appropriate parameters.

livecodeali committed:
[[ AndroidListeners ]] Add Interface Proxy call type for binding string
Add a new java binding string variant - InterfaceProxy. When this type
of binding string is used, the Android LCBInvocationHandler class is
used to create a proxy for the specified listener interface.

livecodeali committed:
[[ AndroidListeners ]] Add docs and release note

runrevmark commented on docs/guides/LiveCode Builder Language Reference.md:
```diff
@@ -575,6 +575,64 @@ includes the specification of function signature, according to the
+targets for java interface callbacks, such as event listeners.
+
+The foreign handler binding to such a function takes a value that should 
+either be a `Handler` or an `Array` - if it is a `Handler`, the handler 
```
I'm not sure 'the handler being called for all callbacks from the specified listener' makes much sense as the callbacks (in which case) all have to have the same signature. Would it be better to say that *if* the interface only has one callback method, then you can use just a handler to bind to it?

livecodeali replied:
Hmm, ok - do you think we should check using reflection on the Java side that there *is* indeed only on callback in this case?

runrevmark replied:
I think so, yes - we need to check the signatures of the handlers anyway using reflection (don't we?) so an extra check to make sure the binding is unambiguous seems to make sense.

runrevmark commented on engine/kernel.gyp:
```diff
@@ -83,12 +83,16 @@
+								# Force the entry point to be included in the output
+								'-Wl,--undefined,Java_com_runrev_android_Engine_doCreate',
+								
+								# Force inclusion of LCB Invocation Handler native callback
```
WE should embelish this comment to explain why - i.e. because the file contains nothing other than the native (JNI) function, which means the object gets discarded by the linker otherwise as nothing in it is statically used.

runrevmark commented on engine/src/mblandroiddc.cpp:
```diff
@@ -2012,6 +2018,29 @@ JNIEXPORT void JNICALL Java_com_runrev_android_Engine_doCreate(JNIEnv *env, jobj
+    MCLog("Got class loader: %p\n", t_class_loader);
+    
+    // The class loader - make sure we hold a global ref.
+    s_android_class_loader = env->NewGlobalRef(t_class_loader);
```
Do we need to delete the global ref on shutdown? (if we are doing so for other such things, that is).

livecodeali replied:
Ah, good catch, thanks

livecodeali committed:
[[ AndroidListeners ]] Add function to check suitability of lcb handlers
- An LCB handler is suitable for a callback if it has the same number of
input parameters as the callback handler and all its parameter typeinfos
are compatible with the JObject typeinfo.

- An LCB handler is suitable for creating a listener proxy if the listener
has only one callback and the handler is suitable for that callback.

- An array of LCB handlers is suitable for creating a listener proxy if
each key in the array is the same as the name of a callback method of the
listener and the handler value for that key is suitable for the callback
with that name.

livecodeali commented:
This needs to throw appropriate errors all over the place.

livecodeali committed:
[[ AndroidListeners ]] GlobalRef incoming jobject callback params
The global refs will be deleted after the callback handler is
invoked.

runrevmark commented on docs/lcb/notes/feature-android_listeners.md:
```diff
@@ -0,0 +1,65 @@
+The foreign handler binding to such a function takes a value that should
+either be a `Handler` or an `Array` - if it is a `Handler`, the specified 
+listener should only have one available callback. If the listener has 
+multiple callbacks, an array can be used to assign handlers to each. Each 
```
Do we need a note about 'overloaded' methods in the interface? I suspect it is a rare-case (as it would be horrendously bad design), but we should perhaps flag it and say it isn't currently supported (unless it is...).

runrevmark commented on engine/src/mblandroiddc.cpp:
```diff
@@ -93,12 +93,12 @@ static jmethodID s_openglview_configure_method = 0;
-static JavaVM *s_java_vm = nil;
+static JavaVM *s_java_vm = nullptr;
 // The JNIEnv for the android UI thread.
 static JNIEnv *s_android_ui_env;
```
This should probably be = nullptr too.

runrevmark commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -1125,38 +1169,258 @@ static bool MCJavaClassNameToPathString(MCNameRef p_class_name, MCStringRef& r_s
+                           MCSTR("com.runrev.android"),
+                           kMCStringOptionCompareExact))
+    {
+#if defined(TARGET_SUBPLATFORM_ANDROID)
```
Perhaps it is worth adding a comment here to describe what we are doing and why (i.e. the whole ClassLoader context thing).

runrevmark commented on docs/guides/LiveCode Builder Language Reference.md:
```diff
@@ -575,6 +575,64 @@ includes the specification of function signature, according to the
+interface. This effectively allows LCB handlers to be registered as the 
+targets for java interface callbacks, such as event listeners.
+
+The foreign handler binding to such a function takes a value that should 
```
See below comment about overloaded interface methods.

runrevmark commented on docs/guides/LiveCode Builder Language Reference.md:
```diff
@@ -575,6 +575,64 @@ includes the specification of function signature, according to the
 [standard rules for forming these](http://journals.ecs.soton.ac.uk/java/tutorial/native1.1/implementing/method.html) 
 when calling the JNI.
 
+The function `interface` may be used on Android to create an interface 
```
We probably need to say something about threading. As it stands, you have to be careful to not to try and use callbacks on any thread other than the UI thread or engine thread or bad things will (in all likelyhood) happen.

runrevmark commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -1464,6 +1733,61 @@ void MCJavaPrivateDestroyObject(MCJavaObjectRef p_object)
+    
+    if (t_handler == nullptr)
+    {
+        MCErrorThrowGenericWithMessage(MCSTR("callback handler not found for listener method %{method}"),
```
I'm a little ponderous over what happens to this error (if it occurs) - this JNI native method is called from Java, in many case without any LCB or LCS related stack-frames. So we are updating the error state without anything 'nearby' to handle it. Indeed, the same is true of an error thrown by what the handler calls.

livecodeali committed:
[[ AndroidListeners ]] Note that overloaded methods in interfaces are not supported

livecodeali committed:
[[ Android ]] Initialise ui env to nullptr

livecodeali committed:
[[ AndroidListeners ]] Explain additional code for finding com.runrev.android classes

livecodeali committed:
[[ AndroidListeners ]] Catch & log errors thrown during listener callback

livecodeali committed:
[[ AndroidListeners ]] Add callback thread warning

runrevmark commented:
@livecode-vulcan review ok 46d66eb

livecode-vulcan commented:
💙 review by @runrevmark ok 46d66ebd2c359e698f25b2bc27533f54aca868cf

livecode-vulcan commented:
😞 test failure 46d66ebd2c359e698f25b2bc27533f54aca868cf

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/359)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/365)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/345)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/288)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/288)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/354)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/359)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/357)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/363)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/357)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/363)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/382)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/353)

livecodeali commented:
@livecode-vulcan retry 46d66eb

livecode-vulcan commented:
😎 test success 46d66ebd2c359e698f25b2bc27533f54aca868cf

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/360)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/366)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/346)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/289)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/289)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/355)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/360)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/358)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/364)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/358)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/364)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/383)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/354)
